### PR TITLE
Adding slack support and integration testing around it into weaviate

### PR DESCRIPTION
**Pull request by knordstrom**

### CODE REVIEW

File: library/weaviate.py [hunk 1]
```diff
@@ -7,222 +7,116 @@
...
+    _client = None
     @property
     def client(self):
-        print("Connecting to " + self.url)
-        return w.connect_to_local(
-            host=self.host,
-            port=self.port,
-        )
-
+        if self._client is None:
```
Comment (knordstrom) [inline library/weaviate.py:20]: Lazy load pattern so only one gets created per wrapper class (Weaviate)

File: library/weaviate.py [hunk 2]
```diff
@@ -7,222 +7,116 @@
...
-        self.port = port
-        self.url = host + ":" + port
-        for schema_entry in schemas:
-            key, schema = schema_entry
-            self.create_schema(schema)  
-            self.schemas[key] = schema 
-
-    def create_schema(self, schema_object) -> None:
+    def __new__(cls, host = '127.0.0.1', port = '8080', schemas: list[(WeaviateSchemas,dict)] = WeaviateSchema.class_objs) -> None:
+        if not hasattr(cls, 'instance'):
```
Comment (knordstrom) [inline library/weaviate.py:33]: Singleton pattern so only one of these wrappers is created per application

File: Makefile [hunk 3]
```diff
@@ -2,7 +2,10 @@ init:
 	poetry install
 
 test:
-	poetry run coverage run -m pytest && poetry run coverage report -m
+	poetry run coverage run -m pytest tests/unit/ && poetry run coverage report -m
```
Comment (knordstrom) [inline Makefile:5]: Tests have been separated into unit and integration folders so that they can be run separately

File: library/neo4j.py [hunk 4]
```diff
@@ -207,13 +207,17 @@ def add_to_db(self, person_list, events_list, attendance_list):
     @staticmethod
     def collate_schedule_response(records):
         def key(record):
-            return record['person.name'] + "||" + record['person.email'] + "||" + record['event.name'] + "||" + str(record['event.start']) + "||" + str(record['event.end'])
+            start = record['event.start'].strftime("%s")
```
Comment (knordstrom) [inline library/neo4j.py:210]: This old format would have been timezone sensitive and failed to collate properly if entries were from people with different iso strings

File: library/weaviate.py [hunk 5]
```diff
@@ -7,222 +7,116 @@
...
-                wvc.config.ReferenceProperty(name="document_id", target_collection="Document"),
-                wvc.config.ReferenceProperty(name="text", target_collection="DocumentText"),
-            ],
-
-            # Specify a vectorizer
-            "vectorizer": True,
-    })
-])
-
+from library.weaviate_schemas import WeaviateSchemas, WeaviateSchema
```
Comment (knordstrom) [inline library/weaviate.py:11]: WeaviateSchemas are now in their own file

File: library/weaviate.py [hunk 6]
```diff
@@ -7,222 +7,116 @@
...
+            if len(post_references) > 0:
+                self.postponded_references[key] = post_references
+                print("Postponding reference creation for ", schema_object['class'], ": ", [property.name for property in post_references])
+
+            self.client.collections.create(schema_object['class'], 
+                    properties = schema_object['properties'], 
+                    references = valid_references,
+                    vectorizer_config = vectorizer                               
+            )                         
+        except w.exceptions.UnexpectedStatusCodeError as e:
```
Comment (knordstrom) [inline library/weaviate.py:65]: This error is very broad so it needed to have a better message such that we can see failure outcomes

File: library/weaviate.py [hunk 7]
```diff
@@ -7,222 +7,119 @@
...
-                    uuid = w.util.generate_uuid5(obj)
+                    uuid = identifier
             )
-
+        return True
+    
+    def upsert_text_vectorized(self, text: str, metaObj: dict, collection_key: WeaviateSchemas) -> bool:
+        collection = self.collection(collection_key)
+        schema_object = WeaviateSchema.class_map[collection_key]
+        reference_keys = [property.name for property in schema_object['references']]
```
Comment (knordstrom) [inline library/weaviate.py]: Identical logic, abstract

File: library/weaviate.py [hunk 8]
```diff
@@ -7,222 +7,119 @@
...
+            self = cls.instance
+            self.host = host
+            self.port = port
+            self.url = host + ":" + port
+            for schema_entry in schemas:
+                key, schema = schema_entry
+                self.create_schema(key, schema)  
+                self.schemas[key] = schema 
+
+            for key, references in self.postponded_references.items():
```
Comment (mshash8) [inline library/weaviate.py]: Would it be possible to create all the schemas first and then populate all the references?

**Pull request merged**